### DISCUSSION TITLE
Tiki Wiki CMS Groupware is making progress on migrating to Smarty 5

### DISCUSSION THREAD

**Discussion by marclaporte** [smarty-php/smarty#920] [General]
Tiki is a big Smarty user: https://doc.tiki.org/Smarty-Templates

Here is the work in progress:

- https://gitlab.com/tikiwiki/tiki/-/merge_requests/3529
- https://gitlab.com/tikiwiki/tiki/-/merge_requests/3897 

We will complete this in January 2024 at the latest. We want it to be part of Tiki 27.0: https://doc.tiki.org/Tiki27

Thanks!

**Comment by wisskid** [comment]
Wow, this is so cool. Thank you for sharing, @marclaporte !

**Comment by wisskid** [comment]
@marclaporte I noticed you have a custom `|ternary` modifier. Did you notice the new [ternary operator](https://smarty-php.github.io/smarty/5.x/designers/language-basic-syntax/language-syntax-operators/#ternary)?

**Reply by rodriguezny** [reply]
@wisskid Yes, we noticed the new [ternary operator](https://smarty-php.github.io/smarty/5.x/designers/language-basic-syntax/language-syntax-operators/#ternary). We opted to implement a custom |ternary modifier to avoid to make changes at many places in our Tiki code where |ternary is used and also for custom templates to avoid causing issues for users that use it in their custom templates. So, for new .tpl files created we will use the new ternary operator, thanks for the add of this new operator.

**Comment by marclaporte** [comment]
Status update: Work is progressing well, and Smarty 5 will be in Tiki 27. However, we have delayed the release by one month.

Over 500 files have been modified: https://gitlab.com/tikiwiki/tiki/-/merge_requests/3529

https://smartydev.tiki.org/ is a test update of https://dev.tiki.org/ using the Smarty 5 branch. https://dev.tiki.org/ is a site started in 2004 so it has a lot of content, and covers a good number of types of contents. Once this works well, the branch will be merged into master. It will be then tested on other test upgrade sites. Once this works well, it will be time to branch 27.x and proceed to the release.

Thank you @wisskid for Smarty 5!

**Comment by marclaporte** [comment]
Smarty 5 has now been merged into Tiki trunk (master):
https://gitlab.com/tikiwiki/tiki/-/merge_requests/3529

Over 800 files have been modified and 277 commits have been squashed into one!

https://smartydev.tiki.org/ will be retired and https://trunkdev.tiki.org/ will be used for testing. Once all is good enough, we will branch 27x and release Tiki 27.0 in the coming weeks: https://doc.tiki.org/Tiki27

Smarty is now at [v5.0.0-rc2](https://github.com/smarty-php/smarty/releases/tag/v5.0.0-rc2). What is a realistic timeline going forward?

Thanks!

**Reply by wisskid** [reply]
I figured I'd wait for your results. If you feel confident that Smarty5 is stable enough, I'll release it asap.

**Reply by marclaporte** [reply]
Ok, we'll test intensely for 7-10 days and we'll report any issues. If we don't see anything major, it's fair to call it 5.0.0 :-)

**Reply by wisskid** [reply]
@marclaporte are there any modifiers that have been left out of the default Extension that you feel should have been included?

**Reply by rodriguezny** [reply]
@wisskid sorry for responding late. The default Extension includes the most important/used modifiers. As you dropped the support for using PHP functions as modifiers, for all PHP functions that were used as modifiers in Tiki and have not been included to the default Extension, we implemented and added them to our custom Extension.
But I think "implode, strpos, preg_match, substr, json_encode, json_decode and urlencode" should have been included to the default Extension, although we already have them included to our custom Extension.
Thanks!

**Reply by wisskid** [reply]
No problem, thank you for your feedback. A couple of points:
- I believe preg_match warrants its own operator to keep the syntax intuitive. I've opened #938 for this.
- The same goes for strpos. I've used it in templates, but it seems awkward. See #937 
- urlencode seems unneeded, since we already have `|escape:url`, or am I missing something? 
- I've personally never used `|json_decode` in a template. Why would you want to do that?

The other ones (implode, substr and json_encode) seem fair. I've added these into our custom extension as well. Let's re-add support for these: #939

**Reply by jonnybradley** [reply]
Hi
We use `json_encode` and `json_decode` often to pass structured data from our search index query results as they're set up to return strings only (so far). It's a slightly nasty workaround but works well _(note to self, need to add some documentation on this, probably [here](https://doc.tiki.org/PluginList-advanced-output-control-block))_
Thanks!

**Comment by marclaporte** [comment]
We will now upgrade to https://github.com/smarty-php/smarty/releases/tag/v5.0.0-rc3

**Reply by marclaporte** [reply]
Done: https://gitlab.com/tikiwiki/tiki/-/merge_requests/4581

We are testing on various sites, and will have feedback soon.

**Reply by marclaporte** [reply]
We are making good progress. Here are Smarty-related commits: https://gitlab.com/tikiwiki/tiki/-/commits/master?search=smarty

**Comment by rodriguezny** [comment]
Hello @wisskid ! I made some changes(https://gitlab.com/tikiwiki/tiki/-/merge_requests/4622) to fix warning message expect the bellow 2 errors that I think should be fixed at your side (smarty).
![Screenshot 2024-03-08 at 15 00 40](https://github.com/smarty-php/smarty/assets/62137963/6e77966b-8406-4169-9549-7ea77c1672f1)
![image](https://github.com/smarty-php/smarty/assets/62137963/7d009262-e158-4948-a890-9d50009b2b08)

**Reply by wisskid** [reply]
Hi @rodriguezny, I'd love to help, but I don't understand what causes these errors. Do you have a reproduction scenario?

**Reply by rodriguezny** [reply]
Hi @wisskid ! In Tiki we have preference/feature called "error_reporting" level which if enabled it shows you PHP error messages thrown during the processing of the page you access and displays them at the botom of the page. so to reproduce this you need to have a Tiki instance.

**Reply by marclaporte** [reply]
To save time, we can set one up for you on a dev server (where do we email credentials?)

On a semi-related note, we use this to centralize error logging:
https://dev.tiki.org/Using-GlitchTip-as-part-of-the-Tiki-development-process

Thanks!

**Reply by wisskid** [reply]
If you can, please open an issue for any bug that you encounter in the [Issues section](https://github.com/smarty-php/smarty/issues). See [Contributing](https://github.com/smarty-php/smarty/blob/293bc20db0f30de54db663ae238c681af90847d9/CONTRIBUTING.md) for some tips on how to write a good bug report.

**Reply by rodriguezny** [reply]
Okay, will do it.

**Reply by rodriguezny** [reply]
@wisskid I also have this error message that seems to thrown in something wrong in our code, but this was not thrown before we upgrade to smarty 5. Any idea of what is wrong in this code ? I found something about the "too many shorthand attribute in the forum" but it didn't help.
`Error: Syntax error in template "file:examples/search/related_tracker_items.tpl"  on line 5 "{if preg_match_all('/(.+?)\:([^\v]+)$/mis', $relationValue, $matches, 2)}" too many shorthand attributes`

The entire content of the {if}{/if} block can be found here : [https://gitlab.com/tikiwiki/tiki/-/blob/master/templates/examples/search/related_tracker_items.tpl](https://gitlab.com/tikiwiki/tiki/-/blob/master/templates/examples/search/related_tracker_items.tpl)

Thanks!

**Reply by rodriguezny** [reply]
I reported an issue [https://github.com/smarty-php/smarty/issues/949](https://github.com/smarty-php/smarty/issues/949)

**Comment by rodriguezny** [comment]
@wisskid can you now release smarty-v5.0.0 ? We are about to release Tiki 27.0 and we want it to be part of. Thanks!

**Reply by wisskid** [reply]
Let me grab a cup of coffee... ☕

**Reply by wisskid** [reply]
https://github.com/smarty-php/smarty/releases/tag/v5.0.0 🍰 🥳 🎉

**Reply by rodriguezny** [reply]
Thanks 👏 🥳  🎉

**Reply by marclaporte** [reply]
Now in Tiki master: https://gitlab.com/tikiwiki/tiki/-/merge_requests/4733

**Comment by marclaporte** [comment]
@rodriguezny @jonnybradley It is now time to start planning for Smarty v6 :-)
https://github.com/smarty-php/smarty/discussions/1200

**Reply by rodriguezny** [reply]
Yes, noted!
